wit-0.3.0-draft/terminal.wit

```diff
@@ -1,25 +1,400 @@
+    /// ensure that if log messages are also being written to the terminal,
+    /// that they not be permitted to be obscured by terminal manipulation.
+    ///
+    /// Output is always postprocessed: U+8 moves the cursor back one column,
+    /// U+9 does the horizontal-tab thing, and U+A performs a line feed and a
+    /// carriage return (`stty opost onlcr -onlret -ocrnl -onocr`).
+    ///
+    /// Terminals may also assign other meanings to other control codes and
+    /// apply other output translations; output is recommended to
+    /// support lowercase (`stty -olcuc`).
+    ///
+    /// By default, terminals are not in full-screen mode or bracketed paste
+    /// mode.
     @since(version = 0.3.0)
-    resource terminal-output;
+    resource terminal-output {
+        /// Return the current number of rows and columns in the terminal.
+        ///
+        /// Not all terminals have a set size, and not all that do know their
+        /// size, so this function may fail if the size cannot be determined.
+        window-size: func() -> result<rows-and-columns>;
+
+        /// Return the current number of columns in the terminal.
+        ///
+        /// Not all terminals have a set size, and not all that do know their
+        /// size, so this function may fail if the size cannot be determined.
+        window-columns: func() -> result<u16>;
+
+        /// Return a `stream` listening for window size changes.
+        ///
+        /// This `stream` produces empty values when the window size changes,
+        /// or when the application is resumed from having been suspended.
+        /// (These correspond to `SIGWINCH` and `SIGCONT` in Unix.)
+        ///
+        /// On implementations which don't support size changes or suspending,
+        /// it never produces any values.
+        size-changes: func() -> stream;
+
+        /// What kinds of colors are supported, and preferred?
+        ///
+        /// This returns a set of flags indicating which families of
+        /// escape sequences for displaying color are supported. The "OSC 4"
+        /// method of detecting color is not supported.
+        ///
+        /// The following additional escape sequences are defined in
+        /// implementations which declare `ansi` color support:
+        ///
+        /// Escape sequences:
+        ///
+        /// | Sequence | Meaning                                              | terminfo  | ti    | tc |
+        /// | -------- | ---------------------------------------------------- | --------- | ----- | -- |
+        /// | `␛[«attrs»m` | Set the foreground and/or background color       | set_attributes | sgr | sa |
+        ///
+        /// `«attrs»` is a semicolon-separated list of decimal-formatted
+        /// integer values from the following list, identifying color
+        /// attributes:
+        ///
+        /// | Foreground | Background | Color |
+        /// | ---- | ---- | ---- |
+        /// | `30` | `40` | Black |
```

Comment:
this should probably include `␛[0m` and `␛[m` to reset colors and other attributes to the default.

Reply:
Good idea, done.